### ISSUE TITLE
Work with array of objects

### ISSUE THREAD

**Issue by NewOldMax** [flitbit/diff#100]
Hi,
In my case, I have some array of objects ordered by `name` key.
This code
````javascript
var lhs = [{id: 1, name: "a"}, {id: 2, name: "b"}, {id: 3, name: "c"}]; // old array, given from backend previous time
var rhs = [{id: 2, name: "b"}, {id: 3, name: "c"}, {id: 1, name: "e"}]; // new array, git from backend now

deepDiff(lhs, rhs);
````
says that I have 6 changes, but in real only 1 object is changed `{id: 1, name: "e"}`. Is there a way to get something like "1 edition in `{id: 1, name: "e"}`"?
I want to use it for a making new items for user, like "hey, this item has been changed". Can I achieve this with your library?

**Comment by tarekz** [comment]
I have the same problem when I delete the first element of an array. It creates n differences for an array of n elements. 
```
var lhs =  ["1", "2", "3", "4"]; // old array, given from backend previous time
var rhs = ["2", "3", "4"];// new array, git from backend now

deepDiff(lhs, rhs);
```

result: 
```
[{
	"kind": "E",
	"path": [0],
	"lhs": "1",
	"rhs": "2"
}, {
	"kind": "E",
	"path": [1],
	"lhs": "2",
	"rhs": "3"
}, {
	"kind": "E",
	"path": [2],
	"lhs": "3",
	"rhs": "4"
}, {
	"kind": "A",
	"index": 3,
	"item": {
		"kind": "D",
		"lhs": "4"
	}
]
````
I just think it should return [{"kind": "A","index": 0,"item": {"kind": "D","lhs": "1"}]

**Comment by vincentsels** [comment]
I think the solution would be to, instead of always comparing objects in both arrays index-by-index in, provide an option to pass a function which specifies how to compare the objects within an array by some kind of unique identifier or hash. Typically this will just be `item => item.id` (as in the first given example) or just be `item => item` (second example).

When such a function is supplied, it is used to find whether this entity exists in the other array. If it does, the objects still needs to be compared because they might have changes.

When the function is not supplied, the current comparison can still be used as fallback (compare objects in both arrays index-by-index).

**Comment by pSchaub** [comment]
@vincentsels I agree with your idea. We need a function to tell the library of to compare objects (e.g. by `item.id` of each object instead of the array key).

**Comment by NewOldMax** [comment]
@pSchaub I'm using something like this
````javascript
import deepDiff from 'deep-diff';

function arrayCompare(arrFirst, arrSecond, propName) {
    const arrDiff = [];
    let fields;
    let obj;
    for (let key2 in arrSecond) {
        obj = arrFirst.filter(o => o[propName] === arrSecond[key2][propName]);
        obj = obj[0];
        if (obj !== undefined) {
            fields = deepDiff(obj, arrSecond[key2]);
        }
        if (fields !== undefined) {
            arrDiff.push({ key: key2, fields });
        }
    }
    if (arrDiff.length) {
        return arrDiff;
    }
    return false;
}

const diff = arrayCompare(firstArr, secondArr, 'id');

````

**Comment by pSchaub** [comment]
btw, this issue is a duplicate, see #37

**Comment by pSchaub** [comment]
@NewOldMax But your solution will not work with nesting (e.g. array > object > array > object > ..), right?

**Comment by NewOldMax** [comment]
Didn't test with nested, in my cases I have only first-level differences

**Comment by pSchaub** [comment]
@NewOldMax I don't see any kind of recursion in your code so this shouldn't work in nested situations. We really need a fix inside the library itself.

**Comment by kenberkeley** [comment]
any progress?

**Comment by pSchaub** [comment]
@kenberkeley We switched to https://github.com/benjamine/jsondiffpatch to solve our nested situation. Maybe this will help you until anybody will implement a way into this package.
